### DISCUSSION TITLE
Jurassic park SNES compatibility?

### DISCUSSION THREAD

**Discussion by ghost** [vgmtrans/vgmtrans#936] [Music formats]
I was wondering if it's okay if compatibility for the SNES Jurassic park could be added as I remember watching my Dad play the game and I have deep fond memories of the game and I would love to get my hands on a Jurassic park SNES Soundfont as well as the Midis. Thank you for taking the time to read this. ^^

VGMTrans version I use is 25/02/2017
 Operating System and version is Windows 10 64 bit
Jurassic Park from snesmusic.org

**Comment by sykhro** [comment]
This game was made by Ocean Software, I doubt anyone has looked into their sound engine at all

**Comment by ghost** [comment]
Thank you very much for replying! I would love to help but I have zero idea on how all this stuff works only surface stuff. Also thank you for adding this to enhancement and good luck with it! ^^

**Comment by loveemu** [comment]
Technical info: SNES Jurassic Park uses a custom sound driver. The code is somewhat based on N-SPC, but the song data format is not compatible with the N-SPC. (In particular, the driver has per-track pattern sequence and different note event format at least.)

I think it will require more work than usual N-SPC variant formats.
https://github.com/loveemu/vgm-disasm/blob/master/snes/Nintendo/Others/Jurassic%20Park.s

**Comment by KungFuFurby** [comment]
I'll go ahead and put in my two cents on this one.

Jurassic Park is the *last* known N-SPC variant version that Ocean used to my knowledge. By tracking by VCMD code starting from the earliest (combined with a length table, though that's not always an indicator), I determined it does qualify as a N-SPC variant.

I'll go ahead and give you a rough ordering by version, earliest to latest:
* Addams Family
* RoboCop 3 *(two different builds with different VCMD versioning)*
* Push-Over
* Lethal Weapon
* Addams Family: Pugsley's Scavenger Hunt
* Cool World *(two different builds, VCMD code came out identical within the scope of what I looked up)*
* Super James Pond II (European version)
* Jurassic Park

Jurassic Park is an unusual one in that code-wise, I ended up using Addams Family: Pugsley's Scavenger Hunt (for internal orderlist handling) and Cool World as a comparison point (Super James Pond II's VCMD code was almost identical, and those changes don't carry over to Jurassic Park, hence me using Cool World).

Here's a list of some of the differences that emerge as we go down the list...
* *All* of them have a unique header format instead of using the phrase format that almost all other N-SPC games use. It's a per-channel orderlist with a few commands in there as well.
* Quantization and Velocity together is only ever used in Addams Family. Every single other game forces quantization to $FF, and instead the extra byte in question is used purely for velocity.
* The instrument data format also gradually gets derivative at least partially because the sample pointers eventually get dynamically set up to some extent. In addition, we end up with two instrument VCMDs starting with RoboCop 3 (with the new one appearing to be the more commonly used one).
* Everything up to Super James Pond 2 generally kept the same internal VCMD IDs with some additions to fill up the entire thing (meaning $FB-$FF are completely used up, with the second build of RoboCop 3 being the first to completely fill up the slots). However...
  * The subroutine VCMD ($EF) is removed starting with the second build of RoboCop 3. It is overwritten by a new VCMD in Jurassic Park.
  * Jurassic Park has no percussion support. As a matter of fact, the entire collection gets pushed back 22 IDs from $E0 to $CA, thus overwriting the entire percussion set in one go. The code for the VCMD in question, $E4 internally (and usually known as $FA in most other variants), still exists, but the pointer goes to the VCMD table instead.
  * Jurassic Park has four more brand new VCMDs added on to the end of the collection. In previous versions, this is an 8-bit overflow, since previously all of the slots were used.
  * One of the VCMDs that was added on (starting with the Addams Family), previously known as $FB but, to Jurassic Park, is $E5, is removed and dummied out with a RET opcode starting with Addams Family: Pugsley's Scavenger Hunt.

**Comment by ghost** [comment]
> 
> 
> I'll go ahead and put in my two cents on this one.
> 
> Jurassic Park is the _last_ known N-SPC variant version that Ocean used to my knowledge. By tracking by VCMD code starting from the earliest (combined with a length table, though that's not always an indicator), I determined it does qualify as a N-SPC variant.
> 
> I'll go ahead and give you a rough ordering by version, earliest to latest:
> 
>     * Addams Family
> 
>     * RoboCop 3 _(two different builds with different VCMD versioning)_
> 
>     * Push-Over
> 
>     * Lethal Weapon
> 
>     * Addams Family: Pugsley's Scavenger Hunt
> 
>     * Cool World _(two different builds, VCMD code came out identical within the scope of what I looked up)_
> 
>     * Super James Pond II (European version)
> 
>     * Jurassic Park
> 
> 
> Jurassic Park is an unusual one in that code-wise, I ended up using Addams Family: Pugsley's Scavenger Hunt (for internal orderlist handling) and Cool World as a comparison point (Super James Pond II's VCMD code was almost identical, and those changes don't carry over to Jurassic Park, hence me using Cool World).
> 
> Here's a list of some of the differences that emerge as we go down the list...
> 
>     * _All_ of them have a unique header format instead of using the phrase format that almost all other N-SPC games use. It's a per-channel orderlist with a few commands in there as well.
> 
>     * Quantization and Velocity together is only ever used in Addams Family. Every single other game forces quantization to $FF, and instead the extra byte in question is used purely for velocity.
> 
>     * The instrument data format also gradually gets derivative at least partially because the sample pointers eventually get dynamically set up to some extent. In addition, we end up with two instrument VCMDs starting with RoboCop 3 (with the new one appearing to be the more commonly used one).
> 
>     * Everything up to Super James Pond 2 generally kept the same internal VCMD IDs with some additions to fill up the entire thing (meaning $FB-$FF are completely used up, with the second build of RoboCop 3 being the first to completely fill up the slots). However...
>       
>       * The subroutine VCMD ($EF) is removed starting with the second build of RoboCop 3. It is overwritten by a new VCMD in Jurassic Park.
>       * Jurassic Park has no percussion support. As a matter of fact, the entire collection gets pushed back 22 IDs from $E0 to $CA, thus overwriting the entire percussion set in one go. The code for the VCMD in question, $E4 internally (and usually known as $FA in most other variants), still exists, but the pointer goes to the VCMD table instead.
>       * Jurassic Park has four more brand new VCMDs added on to the end of the collection. In previous versions, this is an 8-bit overflow, since previously all of the slots were used.
>       * One of the VCMDs that was added on (starting with the Addams Family), previously known as $FB but, to Jurassic Park, is $E5, is removed and dummied out with a RET opcode starting with Addams Family: Pugsley's Scavenger Hunt.

Wow that is so interesting! Thank you for telling me. : D
I used a program called Spc2it and it converted the files pretty well, If you want I could upload the IT files here? : O

**Comment by mikelow** [comment]
We're removing feature requests from Issues. Since there is a lot of good info here, I will convert this into a discussion.
